processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -133,84 +188,37 @@ public int compare(String o1, String o2)
     int valuesOffset = theBuffer.position() + size * Ints.BYTES;
 
     buffer.position(valuesOffset);
-    valueBuffers = Lists.newArrayList(buffer.slice());
+    // Ensure the value buffer's limit equals to capacity.
+    firstValueBuffer = buffer.slice();
+    valueBuffers = new ByteBuffer[]{firstValueBuffer};
     buffer.position(indexOffset);
     headerBuffer = buffer.slice();
-    final ByteBuffer valueBuffer = valueBuffers.get(0);
-    bufferIndexed = new BufferIndexed()
-    {
-      @Override
-      public T get(int index)
-      {
-        checkIndex(index, size);
-
-        final int startOffset;
-        final int endOffset;
-
-        if (index == 0) {
-          startOffset = 4;
-          endOffset = headerBuffer.getInt(0);
-        } else {
-          int headerPosition = (index - 1) * Ints.BYTES;
-          startOffset = headerBuffer.getInt(headerPosition) + Ints.BYTES;
-          endOffset = headerBuffer.getInt(headerPosition + Ints.BYTES);
-        }
-        return _get(valueBuffer.asReadOnlyBuffer(), startOffset, endOffset);
-      }
-    };
   }
 
-  // used for multiple file version, v2.
+
+  /**
+   * Constructor for version two.
+   */
   GenericIndexed(
-      List<ByteBuffer> valueBuffs,
+      ByteBuffer[] valueBuffs,
       ByteBuffer headerBuff,
       ObjectStrategy<T> strategy,
       boolean allowReverseLookup,
       int logBaseTwoOfElementsPerValueFile,
       int numWritten
   )
   {
+    this.versionOne = false;
```

Comment:
I tend to disagree with this.  Using the semantics of the actual objects is more indicative of what is actually happening and more likely to be correct.  By introducing a variable like this, you are introducing a new parameter that can fall out of sync with the rest of the implementation.

IIRC, the places where the version flag are used is during serde, so the performance argument isn't really meaningful.  And given the long-term maintainability of the code, it should stay as it was.